lib/xgu/xgu.h

```diff
@@ -0,0 +1,777 @@
+}
+
+/* ==== Vertex Data Array ==== */
+
+inline
+uint32_t* xgu_set_vertex_data_array_format(uint32_t* p, XguVertexArray index, XguVertexArrayType format,
+                                        uint32_t size, uint32_t stride) {
+    return push_command_parameter(p, NV097_SET_VERTEX_DATA_ARRAY_FORMAT + index*4,
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE, format) |
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_SIZE, size) |
+                                  XGU_MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_STRIDE, stride));
+}
+
+inline
+uint32_t* xgu_set_vertex_data_array_offset(uint32_t* p, XguVertexArray index, const void* data) {
+    return push_command_parameter(p, NV097_SET_VERTEX_DATA_ARRAY_OFFSET + index*4,
+                                  (uint32_t)data/* & 0x03fffff*/);
+}
+
+inline
+uint32_t* xgu_element16(uint32_t* p, const uint16_t* elements, unsigned int count) {
+    assert(count % 2 == 0);
+    p = push_command(p, 0x40000000|NV097_ARRAY_ELEMENT16, count/2);
+    memcpy(p, elements, count * sizeof(uint16_t));
+    p += count / 2;
+    return p;
+}
+
+inline
+uint32_t* xgu_element32(uint32_t* p, const uint32_t* elements, unsigned int count) {
+    p = push_command(p, 0x40000000|NV097_ARRAY_ELEMENT32, count);
+    memcpy(p, elements, count * sizeof(uint32_t));
+    p += count;
+    return p;
+}
+
+inline
+uint32_t* xgu_draw_arrays(uint32_t* p, unsigned int start, unsigned int count) {
+    assert(count>=1);
+    assert(count<=256);
+    return push_command_parameter(p, 0x40000000|NV097_DRAW_ARRAYS,
+                                  XGU_MASK(NV097_DRAW_ARRAYS_COUNT, (count-1)) |
+                                  XGU_MASK(NV097_DRAW_ARRAYS_START_INDEX, start));
+}
+
+/* ==== Alpha/Blend/Cull ==== */
+inline
+uint32_t* xgu_set_alpha_test_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_ALPHA_TEST_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_blend_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_BLEND_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_cull_face_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_CULL_FACE_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_depth_test_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_DEPTH_TEST_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_dither_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_DITHER_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_lighting_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_LIGHTING_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_stencil_test_enable(uint32_t* p, bool enabled) {
+    return push_command_boolean(p, NV097_SET_STENCIL_TEST_ENABLE, enabled);
+}
+
+inline
+uint32_t* xgu_set_alpha_func(uint32_t* p, uint8_t func) {
```

Comment:
func should probably be an enum?

Reply:
Not improbable, `nv_regs.h` doesn't give much of a hint on what the possible values are though.

Reply:
check what XQEMU does? clearly it must handle this

inb4 XQEMU doesn't have `alpha_func`

Isn't nv_regs the pbkit header? we have the XQEMU header in nv2a_regs.h I think?

Reply:
# This discussion can be removed